server/src/test/java/io/crate/execution/dml/upsert/TransportShardUpsertActionTest.java

```diff
@@ -363,4 +377,118 @@ public void testItemsWithoutSourceAreSkippedOnReplicaOperationBWC() throws Excep
+        // id is a primary key, column "b" is not in targets.
+        ShardUpsertRequest request = new ShardUpsertRequest.Builder(
+            DUMMY_SESSION_INFO,
+            TimeValue.timeValueSeconds(30),
+            DuplicateKeyAction.UPDATE_OR_FAIL,
+            false,
+            new String[] {updatedRef.column().name()},
+            new SimpleReference[]{ID_REF, updatedRef},
+            null,
+            UUID.randomUUID()
+        ).newRequest(shardId);
+
+
+        var item = ShardUpsertRequest.Item.forInsert(
+            "1",
+            List.of("1"),
+            Translog.UNSET_AUTO_GENERATED_TIMESTAMP,
+            new Object[]{(short) 1, 2},
+            new Symbol[]{Literal.of(3)}
+        );
+        request.add(1, item);
+
+
+        TransportShardUpsertAction transportShardUpsertAction = new TransportShardUpsertAction(
+            Settings.EMPTY,
+            mock(ThreadPool.class),
+            clusterService,
+            MockTransportService.createNewService(Settings.EMPTY, VersionUtils.randomVersion(random()), THREAD_POOL, nettyBootstrap, clusterService.getClusterSettings()),
+            mock(TransportAddColumnAction.class),
+            mock(TasksService.class),
+            indicesService,
+            mock(ShardStateAction.class),
+            createNodeContext(),
+            schemas
+        );
+
+        TransportWriteAction.WritePrimaryResult<ShardUpsertRequest, ShardResponse> result =
+            transportShardUpsertAction.processRequestItems(indexShard, request, new AtomicBoolean(false));
+
+        // Imitate what happened on bug:
+        // primary write was successful but replica request was invalid (had mismatching refs/values).
+        // Since streaming of request values based on corresponding target column type,
+        // streaming of replica request could fail on mismatching (column, value) pairs.
+        assertThat(result.finalResponseIfSuccessful.failure()).isNull();
+
+        // processRequestItemsOnReplica was never hit due to streaming failures.
+        // Imitate streaming of the replica request and make sure it doesn't throw an error.
+        ShardUpsertRequest replicaRequest = result.replicaRequest();
+        TransportReplicationAction.ConcreteShardRequest<ShardUpsertRequest> shardRequest =
+            new TransportReplicationAction.ConcreteShardRequest<>(replicaRequest, "dummy", 0L);
+
+        BytesStreamOutput out = new BytesStreamOutput();
+        // Used to fail with java.lang.ClassCastException: class java.lang.Long cannot be cast to class java.lang.String
+        shardRequest.writeTo(out);
+
+        StreamInput in = out.bytes().streamInput();
+        TransportReplicationAction.ConcreteShardRequest<ShardUpsertRequest> shardRequest2 =
+            new TransportReplicationAction.ConcreteShardRequest<>(in, ShardUpsertRequest::new);
+
+        assertThat(shardRequest.getRequest()).isEqualTo(shardRequest2.getRequest());
```

Comment:
Shouldn't we just assert on the columns set on the `replicaRequest` instead? Afaik that is the issue we fix, streaming errors are just a consequence of it.

Reply:
Thanks, that revealed that I'm testing only insert scenario, ie unused column is not part of the replica request currently as with the current setup be are basically using "inserting" indexer.

It's not a bug but we need to target conflict scenario. ie use `TestingTransportShardUpsertAction` which imitates conflict. `PKLookupOperation` needs to be mocked, will probably refactor `TransportShardUpsertAction` to make PKLookup injectable/testable instead of mocking out the whole PKLookup.